btb/basic_btb/basic_btb.cc

```diff
@@ -0,0 +1,239 @@
+    basic_btb_indirect[cpu][i] = 0;
+  }
+  basic_btb_conditional_history[cpu] = 0;
+
+  for (uint32_t i = 0; i < BASIC_BTB_RAS_SIZE; i++) {
+    basic_btb_ras[cpu][i] = 0;
+  }
+  basic_btb_ras_index[cpu] = 0;
+  for (uint32_t i=0; i<BASIC_BTB_CALL_INSTR_SIZE_TRACKERS; i++) {
+    basic_btb_call_instr_sizes[cpu][i] = 4;
+  }
+}
+
+uint64_t O3_CPU::btb_prediction(uint64_t ip, uint8_t branch_type,
+                                uint8_t &always_taken) {
+  if (branch_type != BRANCH_CONDITIONAL) {
+    always_taken = true;
+  }
+
+  if ((branch_type == BRANCH_DIRECT_CALL) ||
+      (branch_type == BRANCH_INDIRECT_CALL)) {
+    // add something to the RAS
+    push_basic_btb_ras(cpu, ip);
+  }
+
+  if (branch_type == BRANCH_RETURN) {
+    // peek at the top of the RAS
+    uint64_t target = peek_basic_btb_ras(cpu);
+    // and adjust for the size of the call instr
+    target += basic_btb_get_call_size(cpu, target);
+
+    return target;
+  } else if ((branch_type == BRANCH_INDIRECT) ||
+             (branch_type == BRANCH_INDIRECT_CALL)) {
+    return basic_btb_indirect[cpu][basic_btb_indirect_hash(cpu, ip)];
+  } else {
+    // use BTB for all other branches + direct calls
+    auto btb_entry = basic_btb_find_entry(cpu, ip);
+
+    if (btb_entry == NULL) {
+      // no prediction for this IP
+      always_taken = 1;
+      return 0;
+    }
+
+    always_taken = btb_entry->always_taken;
+    basic_btb_update_lru(cpu, btb_entry);
+
+    return btb_entry->target;
+  }
+
+  return 0;
+}
+
+void O3_CPU::update_btb(uint64_t ip, uint64_t branch_target, uint8_t taken,
+                        uint8_t branch_type) {
+  // updates for indirect branches
+  if ((branch_type == BRANCH_INDIRECT) ||
+      (branch_type == BRANCH_INDIRECT_CALL)) {
+    basic_btb_indirect[cpu][basic_btb_indirect_hash(cpu, ip)] = branch_target;
+  }
+  if (branch_type == BRANCH_CONDITIONAL) {
+    basic_btb_conditional_history[cpu] <<= 1;
+    if (taken) {
+      basic_btb_conditional_history[cpu] |= 1;
+    }
+  }
+
+  if (branch_type == BRANCH_RETURN) {
+    // recalibrate call-return offset
+    // if our return prediction got us into the right ball park, but not the
+    // exactly correct byte target, then adjust our call instr size tracker
+    uint64_t call_ip = pop_basic_btb_ras(cpu);
+    uint64_t estimated_call_instr_size = basic_btb_abs_addr_dist(call_ip, branch_target);
+    if (estimated_call_instr_size <= 10) {
+      basic_btb_call_instr_sizes[cpu][basic_btb_call_size_tracker_hash(call_ip)] = estimated_call_instr_size;
+    }
+  } else if ((branch_type != BRANCH_INDIRECT) &&
+	     (branch_type != BRANCH_INDIRECT_CALL)) {
+    // use BTB
+    auto btb_entry = basic_btb_find_entry(cpu, ip);
+
+    if (btb_entry == NULL) {
+      if ((branch_target != 0) && taken) {
+        // no prediction for this entry so far, so allocate one
+        uint64_t set = basic_btb_set_index(ip);
+        auto repl_entry = basic_btb_get_lru_entry(cpu, set);
```

Comment:
I have an easy way to compare and find LRU, but it's currently limited to src/base_replacement.cc. Would you want me to insert it in here? I'm a big fan of code reuse.

Reply:
I actually think that these user-definable modules should be totally self-contained, so that it will be easier to maintain backward and forward compatibility in the future.  I think that's what we've done so far with all the others, like prefetchers and branch predictors.